### PR TITLE
Homework 3

### PR DESCRIPTION
**Pull request by Kuzeikind**
Test refactored and passing.

### CODE REVIEW

File: src/main/java/com/epam/tc/hw3/components/leftside/LeftSideMenu.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+import org.openqa.selenium.support.PageFactory;
+
+public class LeftSideMenu extends AbstractMenu {
+
+    @FindBy(xpath = LeftSideMenuLocators.MENU_ITEM_XPATH)
+    protected List<WebElement> menuItems;
+    protected ServicesMenu servicesMenu;
```
Comment (akulebyakin) [inline src/main/java/com/epam/tc/hw3/components/leftside/LeftSideMenu.java:17]: Why not private?

**Review by akulebyakin — CHANGES_REQUESTED** [review]
From my point of view pages hierarchy is way too complicated for the application (I hope this diagram I created is correct)
image

- CheckBoxSection and RadioSection should't be inherited from AbstractMenu
- For pages there can be only one-layer of inheritance. I'd be much easier to support and extend

image

Please review the comments.

File: src/main/java/com/epam/tc/hw3/AbstractPage.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package com.epam.tc.hw3;
+
+import java.time.Duration;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.support.ui.WebDriverWait;
+
+public class AbstractPage {
```
[the remark this answers is not in the document]
Comment (Kuzeikind) [reply src/main/java/com/epam/tc/hw3/AbstractPage.java]: Made the class abstract. Was a mistake.

File: src/main/java/com/epam/tc/hw3/BasePage.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+                   frame -> {
+                       final boolean found;
+                       driver.switchTo().frame(frame);
+                       found = wait.until(
+                           ExpectedConditions.presenceOfAllElementsLocatedBy(HomePageLocators.FRAME_BUTTON)
+                       ).size() > 0;
+                       driver.switchTo().parentFrame();
+                       return found;
+                   })
+               .findAny();
```
[the remark this answers is not in the document]
Comment (Kuzeikind) [reply src/main/java/com/epam/tc/hw3/BasePage.java]: Changed to list.isEmpty().

File: src/main/java/com/epam/tc/hw3/driverutil/DriverServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    @Override
+    public void setupDriver() {
+        DriverManagerType driverManagerType = DriverManagerType.valueOf(
+            properties.getProperty("driver").toUpperCase());
+        WebDriverManager webDriverManager = WebDriverManager.getInstance(driverManagerType);
+        webDriverManager.setup();
+    }
+
+    public WebDriver getDriver() {
+        WebDriver driver = new ChromeDriver();
```
[the remark this answers is not in the document]
Comment (Kuzeikind) [reply src/main/java/com/epam/tc/hw3/driverutil/DriverServiceImpl.java]: Reworked the class to provide different browsers as specified in the properties.

File: src/test/java/com/epam/tc/hw3/BaseTest.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+    @BeforeMethod
+    public void setAssertions() {
+        assertions = new SoftAssert();
+    }
+
+    @AfterMethod
+    public void resetAssertions() {
+        assertions = null;
+    }
```
[the remark this answers is not in the document]
Comment (Kuzeikind) [reply src/test/java/com/epam/tc/hw3/BaseTest.java]: Rewrote methods to exclude multiple BeforeMethod's and BeforeClass'es.

File: src/test/java/com/epam/tc/hw3/BaseTest.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
+package com.epam.tc.hw3;
+
+import com.epam.tc.hw2.util.PageData;
+import com.epam.tc.hw2.util.TestUtil;
```
[the remark this answers is not in the document]
Comment (Kuzeikind) [reply src/test/java/com/epam/tc/hw3/BaseTest.java]: Corrected the mistake.

File: src/main/java/com/epam/tc/hw3/AbstractPage.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Kuzeikind) [reply src/main/java/com/epam/tc/hw3/AbstractPage.java]: > Made the class abstract. Was a mistake.

Removed the class totally.

File: src/main/java/com/epam/tc/hw3/BasePage.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+import com.epam.tc.hw3.locators.HomePageLocators;
+import java.util.List;
+import java.util.Optional;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+import org.openqa.selenium.support.PageFactory;
+import org.openqa.selenium.support.ui.ExpectedConditions;
+
+public class BasePage extends AbstractPage {
```
[the remark this answers is not in the document]
Comment (Kuzeikind) [reply src/main/java/com/epam/tc/hw3/BasePage.java:15]: Reworked the page classes hierarchy, reducing the depth of inheritance.

File: src/main/java/com/epam/tc/hw3/HomePage.java [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.support.PageFactory;
+
+public class HomePage extends BasePage {
+
+    public HomePage(WebDriver driver) {
+        super(driver);
+        PageFactory.initElements(driver, this);
+    }
```
[the remark this answers is not in the document]
Comment (Kuzeikind) [reply src/main/java/com/epam/tc/hw3/HomePage.java:11]: Shouldn't be an issue with the new pages hierarchy.

File: src/main/java/com/epam/tc/hw3/components/BenefitsSection.java [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+import org.openqa.selenium.support.PageFactory;
+
+public class BenefitsSection extends AbstractComponent {
+
+    @FindBy(xpath = HomePageLocators.BENEFIT_ICON_WRAPPER_XPATH)
+    protected List<WebElement> benefitIcons;
+
+    public BenefitsSection(WebDriver driver) {
+        super(driver);
+        PageFactory.initElements(driver, this);
```
[the remark this answers is not in the document]
Comment (Kuzeikind) [reply src/main/java/com/epam/tc/hw3/components/BenefitsSection.java:19]: Disagree. Each component has its own unique web elements and we need to initialise them in each class separately.

File: src/test/java/com/epam/tc/hw3/BaseTest.java [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+    protected WebDriverWait driverWait(long seconds) {
+        return new WebDriverWait(driver, Duration.ofSeconds(seconds));
+    }
+
+    protected void assertBasePageLoaded(WebDriver driver) {
+
+        BasePage basePage = new BasePage(driver);
+        assertions.assertEquals(basePage.getTitle(), PageData.homePageTitle);
+    }
```
[the remark this answers is not in the document]
Comment (Kuzeikind) [reply src/test/java/com/epam/tc/hw3/BaseTest.java:89]: HomePage is now the page loaded for the URL. BasePage is now the abstract superclass for all pages.

File: src/main/java/com/epam/tc/hw3/components/leftside/LeftSideMenu.java (hunk 1 above)
Comment (Kuzeikind) [reply src/main/java/com/epam/tc/hw3/components/leftside/LeftSideMenu.java]: Changed to private.

**Review by akulebyakin — APPROVED** [review]
Approved
Great job!

**Pull request merged**